File: src/features/students/studentInsights.ts
import { format, subMonths } from "date-fns";
import type { AttendanceRecord, PaymentRecord, Student, TrendPoint } from "../../types/app";

export interface StudentMonthOption {
  key: string;
  label: string;
}

const RECENT_MONTH_COUNT = 6;

function monthStartFromKey(monthKey: string) {
  return new Date(`${monthKey}-01T00:00:00`);
}

function attendanceMonthKey(record: AttendanceRecord) {
  return record.attendance_date.slice(0, 7);
}

function paymentMonthKey(record: PaymentRecord) {
  return `${record.payment_year}-${String(record.payment_month).padStart(2, "0")}`;
}

function recentMonthKeys(count = RECENT_MONTH_COUNT) {
  return Array.from({ length: count }, (_, index) =>
    format(subMonths(new Date(), count - index - 1), "yyyy-MM"),
  );
}

export function currentMonthKey() {
  return format(new Date(), "yyyy-MM");
}

export function formatMonthKeyLabel(monthKey: string, pattern = "MMMM yyyy") {
  return format(monthStartFromKey(monthKey), pattern);
}

export function buildStudentMonthOptions(
  student: Student,
  attendance: AttendanceRecord[],
  payments: PaymentRecord[],
): StudentMonthOption[] {
  const keys = new Set<string>(recentMonthKeys());

  keys.add(currentMonthKey());
  keys.add(student.joined_date.slice(0, 7));

  attendance.forEach((record) => keys.add(attendanceMonthKey(record)));
  payments.forEach((record) => keys.add(paymentMonthKey(record)));

  return Array.from(keys)
    .sort((left, right) => right.localeCompare(left))
    .map((key) => ({
      key,
      label: formatMonthKeyLabel(key),
    }));
}

export function studentAttendanceTrend(attendance: AttendanceRecord[]): TrendPoint[] {
  return recentMonthKeys().map((monthKey) => ({
    label: formatMonthKeyLabel(monthKey, "MMM yy"),
    value: attendance.filter(
      (record) => attendanceMonthKey(record) === monthKey && record.status !== "absent",
    ).length,
  }));
}

export function studentPaymentTrend(payments: PaymentRecord[]): TrendPoint[] {
  return recentMonthKeys().map((monthKey) => ({
    label: formatMonthKeyLabel(monthKey, "MMM yy"),
    value: payments.some((record) => paymentMonthKey(record) === monthKey && record.paid) ? 1 : 0,
  }));
}

export function attendanceForMonth(attendance: AttendanceRecord[], monthKey: string) {
  return [...attendance]
    .filter((record) => attendanceMonthKey(record) === monthKey)
    .sort((left, right) => right.attendance_date.localeCompare(left.attendance_date));
}

export function paymentsForMonth(payments: PaymentRecord[], monthKey: string) {
  return [...payments]
    .filter((record) => paymentMonthKey(record) === monthKey)
    .sort((left, right) => {
      const leftKey = `${left.payment_year}-${String(left.payment_month).padStart(2, "0")}`;
      const rightKey = `${right.payment_year}-${String(right.payment_month).padStart(2, "0")}`;
      return rightKey.localeCompare(leftKey);
    });
}

export function countAttendedClasses(attendance: AttendanceRecord[]) {
  return attendance.filter((record) => record.status !== "absent").length;
}

File: apps/android/app/src/main/java/com/udchemistry/mobile/util/StudentInsights.kt
package com.udchemistry.mobile.util

import com.udchemistry.mobile.model.AttendanceRecord
import com.udchemistry.mobile.model.PaymentRecord
import com.udchemistry.mobile.model.Student
import com.udchemistry.mobile.model.TrendPoint
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter

data class StudentMonthOption(
    val key: String,
    val label: String,
)

private val monthKeyFormatter = DateTimeFormatter.ofPattern("yyyy-MM")
private val monthLabelFormatter = DateTimeFormatter.ofPattern("MMMM yyyy")
private val monthChartFormatter = DateTimeFormatter.ofPattern("MMM yy")

private fun recentMonths(count: Int = 6): List<YearMonth> {
    val current = YearMonth.now()
    return (count - 1 downTo 0).map { offset -> current.minusMonths(offset.toLong()) }
}

private fun attendanceMonthKey(record: AttendanceRecord): String = record.attendanceDate.take(7)

private fun paymentMonthKey(record: PaymentRecord): String {
    return "%04d-%02d".format(record.paymentYear, record.paymentMonth)
}

fun currentMonthKey(): String = YearMonth.now().format(monthKeyFormatter)

fun buildStudentMonthOptions(
    student: Student,
    attendance: List<AttendanceRecord>,
    payments: List<PaymentRecord>,
): List<StudentMonthOption> {
    val monthKeys = linkedSetOf<String>()
    recentMonths().forEach { monthKeys += it.format(monthKeyFormatter) }
    monthKeys += currentMonthKey()
    monthKeys += student.joinedDate.take(7)
    attendance.forEach { monthKeys += attendanceMonthKey(it) }
    payments.forEach { monthKeys += paymentMonthKey(it) }
    return monthKeys
        .sortedDescending()
        .map { StudentMonthOption(it, formatMonthKeyLabel(it)) }
}

fun buildStudentAttendanceTrend(attendance: List<AttendanceRecord>): List<TrendPoint> {
    return recentMonths().map { month ->
        val monthKey = month.format(monthKeyFormatter)
        TrendPoint(
            label = month.format(monthChartFormatter),
            value = attendance.count {
                attendanceMonthKey(it) == monthKey && it.status.lowercase() != "absent"
            }.toDouble(),
        )
    }
}

fun buildStudentPaymentTrend(payments: List<PaymentRecord>): List<TrendPoint> {
    return recentMonths().map { month ->
        val monthKey = month.format(monthKeyFormatter)
        TrendPoint(
            label = month.format(monthChartFormatter),
            value = if (payments.any { paymentMonthKey(it) == monthKey && it.paid }) 1.0 else 0.0,
        )
    }
}

fun attendanceForMonth(attendance: List<AttendanceRecord>, monthKey: String): List<AttendanceRecord> {
    return attendance
        .filter { attendanceMonthKey(it) == monthKey }
        .sortedByDescending { it.attendanceDate }
}

fun paymentsForMonth(payments: List<PaymentRecord>, monthKey: String): List<PaymentRecord> {
    return payments
        .filter { paymentMonthKey(it) == monthKey }
        .sortedWith(compareByDescending<PaymentRecord> { it.paymentYear }.thenByDescending { it.paymentMonth })
}

fun countAttendedClasses(attendance: List<AttendanceRecord>): Int {
    return attendance.count { it.status.lowercase() != "absent" }
}

fun formatMonthKeyLabel(monthKey: String): String {
    return runCatching {
        YearMonth.parse(monthKey, monthKeyFormatter).format(monthLabelFormatter)
    }.getOrDefault(monthKey)
}

fun formatShortAttendanceDate(value: String): String {
    return runCatching {
        LocalDate.parse(value).format(DateTimeFormatter.ofPattern("dd MMM"))
    }.getOrDefault(value)
}
